### ISSUE TITLE
v0

### ISSUE THREAD

**Issue by rom1504** [iejMac/video2dataset#18]
https://docs.google.com/document/d/1_TD2KQLkEegszq4Eip568fc6cWnh9h0Jqj4Lc88t9Y0/edit#bookmark=id.v1npsrmmerim

Goals:
1000 hours of video/audio processed
3 sources supported
3 formats supported
Used by 2 training codebase

Features:
Input: files and streams
Output: webdataset
Transform: simple subsampling
Basic documentation
Proper pip packaging

Speed:
Basic subsampling, 1h processed in 1h of cpu core

https://github.com/iejMac/video2dataset/issues?q=is%3Aissue+is%3Aopen+label%3Av0

**Comment by rom1504** [comment]
Priority queue:

* https://github.com/iejMac/video2dataset/issues/25

**Comment by iejMac** [comment]
@rom1504 
Making a summary/list of what needs to be done pre-release. Going over the code as it's executed and noting my thoughts:

## main.py
- [dict as input type for encode_formats](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/main.py#L49). Works when calling video2dataset from python but not sure how this works with fire/CLI. Any thoughts?

## worker.py
- [make this cleaner](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/worker.py#L160). ifs are ugly and doesn't get the point across, point is that the video already has the audio so you want something like ```bytes_downloaded += max(streams.get("video", 0), streams.get("audio", 0))```
- [clipping subsampler should take encode_formats as init param](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/worker.py#L169)
- [we can make this nicer](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/worker.py#L167) by doing something like ```broadcast_subsampler = clipping_sub if "clips" in whatever else noop_sub``` and then just call that with streams and meta
- [idea to get rid of this](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/worker.py#L174) and listing out all subsamples, as we add more we shouldn't have to add another if statement in the worker loop. The idea is to initialize worker with a list of non-None f"{modality}_subsamplers" list and then just iterate over all of those. The reason the attribute would be called f"{modality}_subsamplers" is because instead of checking if we have "video" in streams we can just iterate over all the modalities in streams and retrieve the ```eval(f"self.{modality}_subsamplers")
-  [format_type isn't argument to writer](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/worker.py#L195)

## data_reader.py
- [rename and rethink Mp4Downloader](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L105). Maybe something like ```WebFileDownloader``` or whatever. Should be very similar process for any .mp4, .mp3, .webm, .wav, or whatever extension links. Maybe we can add some is_web_link helper function and you pass the url into and then if that returns true you go to WebFileDownloader that determines what to do with bytes based on encode_formats. i.e. if you have audio then bytes are audio, if you have video,audio then the url is video but you need to extract the audio etc.
- [let's commit to output dicts](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L129) kind of like with streams and encode_formats, we have this everywhere and makes the code more readable imo
- [sample_rate shouldn't be a key in encode_formats](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L112), it should be a parameter to main which gets used in some audio subsampler or whatever
- [video2audio is kind of already doing fusing without an explicit audio subsampler](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L16), I'll put this in an audio subsampler and we can come back to this when fusing. Although this isn't great fusing tbh. Also add a testing function for this in test_subsamplers.py
- [also video2audio needs better param names](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L11)
- [remove this](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L202), and rename things so it's "audio_file", "video_file". Let's keep naming conventions involving modalities consistent to stay open to tricks like ```eval(f"{modality}_something")``` and such. Also we can reduce some lines in the VideoDataReader class with those tricks
- [audio tempfile isn't deleted](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L227)
- [even though for now it is yt_meta_dict let's just call it meta_dict](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L230)
- [YtDlpDownloader needs editing](https://github.com/iejMac/video2dataset/blob/203e4f8dd5845124eee3bc3ac54a7bd282745d5d/video2dataset/data_reader.py#L100), better format string, why m4a? why not the extension passed in via encode_formats? also it does subsampling in the downloader which we don't want yet. Also you can probably do something similar as we do for resolution i.e. get closest possible higher sample rate through passing the correct format string. [Also I'm not sure we can just replace like this](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_reader.py#L168) and finally I need to make sure video is downloaded separately from audio and audio separately from video, then we can download separately and also edit that max() line above to just sum the bytes then it will be exactly right. Some useful links for me later:https://write.corbpie.com/downloading-youtube-videos-as-audio-with-yt-dlp/, https://superuser.com/questions/1266162/youtube-dl-set-sample-rate-on-mp3
- in general data reader needs to generalize to other video formats than mp4 (right now it's overfit to that)

## data_writer.py
- [think about this](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/data_writer.py#L107), do we need to be iterating over encode_formats? maybe it's enough to just iterate over streams and write nothing for cases where a modality isn't present in streams. When does this happen? Do we want to write an empty meta then?

## subsamplers/clipping_subsampler.py
- [same as with noop_subsampler](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/video2dataset/subsamplers/clipping_subsampler.py#L34), encode_formats should be init param not __call__ param
- need to check that audio clipping works as intended though i.e. lines up with video and correct number of clips correctly ID'd etc.

## tests/test_downloaders.py
- [doesn't test if audio is downloaded at all, needs to be added](https://github.com/iejMac/video2dataset/blob/4fe9c5917bdb8b36ec84d3fe771f29520f7bc279/tests/test_downloaders.py#L17)

## tests/test_audio.py
- should be renamed to test_reader.py and we should actually test the reader
- actually pretty good just needs to be adapted a bit and add more parameters to test if the video is reading properly etc. we should also test that the correct error_messages are being returned (or exceptions being thrown if we decide to merge that one PR)

## README.md
- examples getting too long and unintuitive, we should just make more things in examples directory
- specifically let's show how to use encode_formats and other params like that
- also let's add a tutorial for how to run this with distributed=spark, I think that's not obvious but very useful.

besides the above cleanup there's 2 more PR's to get merged:
- https://github.com/iejMac/video2dataset/pull/91 - improves subtitle support and fixes a few things, 100% needs to get merged
- https://github.com/iejMac/video2dataset/pull/92 - I think this is worth trying and considering

# v1 ideas
While going through the code I had some ideas for v1:
- if encode_formats has both video and audio perhaps we should try to do most of the pipeline with video and audio in 1 mp4 byte stream instead of separate video and audio streams so subsampling such as clipping can be done together and then we can split it up instead of the other way

**Comment by iejMac** [comment]
Delaying, we need to get some successful use case of data from this repo. Either SVD or VideoCLIP or whatever.

**Comment by rom1504** [comment]
I would call this done

codebase are using this

**Comment by iejMac** [comment]
yeah sure, maybe it would be good to update pip package if this is the case
